develop-docs/sdk/telemetry/logs.mdx

```diff
@@ -0,0 +1,381 @@
+
+Sentry.logger.emit(
+  level='info',
+  message='Adding item for user',
+  attributes={
+    'item_id': item_id,
+    'user_id': user_id,
+    extra: 123
+  },
+)
+```
+
+#### PHP
+
+```php
+use function Sentry\logger;
+
+logger()->info('Adding item %s for user %s', [$item_id, $user_id], ['extra' => 123]);
+
+logger()->emit(
+  level: LogLevel::info(),
+  message: 'Adding item for user',
+  attributes: [
+    'item_id': item_id,
+    'user_id': user_id,
+    'extra': 123
+  ],
+);
+```
+
+#### Java
+
+```java
+import io.sentry.Sentry;
+
+// example with MessageFormat based string template
+Sentry.logger().info("Adding item {0,string} for user {1,string}", item_id, user_id);
+
+Sentry.logger().emit(log -> {
+  log.setLevel("info");
+  log.setMessage("Adding item for user");
+  log.setAttribute("item_id", item_id);
+  log.setAttribute("user_id", user_id);
+  log.setAttribute("extra", 123);
+});
+
+// Java - Builder Pattern, where setters return builder for method chaining
+Sentry.Logger.emit(log -> {
+  log.setLevel("info")
+  .setMessage("Adding item for user")
+  .setAttribute("item_id", item_id)
+  .setAttribute("user_id", user_id)
+  .setAttribute("extra", 123)
+});
+
+// Kotlin
+Sentry.Logger.emit("Adding item for user") { // fun emit(block: LogItem.() -> Unit)
+  setLevel("info")
+  setAttribute("item_id", item_id)
+  setAttribute("user_id", user_id)
+  setAttribute("extra", 123)
+}
+```
+
+#### Apple
+
+```swift
+// Swift
+SentrySDK.logger
+  .info(message: "Adding item %s for user %s",
+    params: [itemId, userId],
+    attributes: ["extra": @"123"]
+  )
+
+// Objective-C
+[SentrySDK.logger
+    emit :@"Adding item for user"
+    level: @"info"
+    attributes: @{ @"extra" : @"123", @"item_id" : item_id, @"user_id" : user_id }
+];
+```
+
+#### Threading and Concurrency Considerations
+
+Both the `Sentry.logger.X` and `Sentry.logger.emit` methods are fire-and-forget (have no return value). This means that the SDK can choose to run these methods on a separate thread or queue them up for processing later. This includes evaluating the string template, and running any internal hooks (like `beforeSendLog`). The SDK should ensure that the logs are sent in the order they are received.
+
+## SDK Behavior
+
+### Implementation
+
+If `debug` is set to `true` in SDK init, calls to the Sentry logger API should also print to the console with the appropriate log level. This will help debugging logging setups.
+
+Logs should be buffered before being sent. SDKs should keep a buffer of logs on the client (so you can have logs from multiple traces in the buffer) that flushes out based on some kind of condition. A recommended condition is if the buffer length exceeds 25 items, or if 5 seconds have passed. In the future we can use a [weight-tracking based approach](https://github.com/getsentry/rfcs/blob/main/text/0126-sdk-spans-aggregator.md#weight) to determine when to flush logs.
```

Comment:
For SDK with an offline cache, we will run into cache limits when users log extensively. This then can lead to dropping more important envelopes containing sessions updates, error, or crashes when the cache is full and completely filling it with logs only. We already discussed in the past that maybe we need to split these caches or come up with some kind of priority queue, so logs get deleted before crash events, for example. Currently, the default cache limit on most mobile SDKs is 100 envelopes.

I'm not sure if we already need to address this now and maybe we can skip it for LA.

Reply:
We can skip for LA. One we get client outcomes data we can come back and make a more formal decision.